File: semilio-core/src/main/java/com/example/semilio/product/ProductService.java
package com.example.semilio.product;

import com.example.semilio.product.request.ProductRequestDTO;
import com.example.semilio.product.request.ProductSearchCriteriaRequest;
import com.example.semilio.product.response.ProductCardResponse;
import com.example.semilio.product.response.ProductDetailDTO;
import com.example.semilio.product.response.ProductSummaryDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

@Service
public interface ProductService {

    ProductCardResponse createProduct(ProductRequestDTO productDTO,
                                      List<MultipartFile> images,
                                      Authentication principal);

    ProductSummaryDTO updateProduct(String productId,
                                    ProductRequestDTO productDTO,
                                    List<MultipartFile> newFiles,
                                    Authentication principal);

    ProductDetailDTO getProductById(String productId);

    void deleteProduct(String productId, Authentication principal);

    Page<ProductCardResponse> getUserProducts(Authentication principal,
                                              Pageable pageable);

    void changeVisibility(String productId, Authentication principal);

    Page<ProductCardResponse> getFeaturedProducts(Authentication principal,
                                                  String effectiveSeed,
                                                  Pageable pageable);

    Page<ProductCardResponse> search(ProductSearchCriteriaRequest criteria,
                                     Pageable pageable,
                                     Authentication auth);

    Page<ProductCardResponse> getSellerProducts(Authentication principal,
                                                String sellerId,
                                                Pageable pageable);

    void addViewAsync(String productId);

    ProductDetailDTO getProductBySlug(String slug);
}

File: similio-web/src/app/core/service/product.service.ts
import {Injectable} from '@angular/core';
import {IProductCard} from '../models/product/i-product-card';
import {IProduct} from '../models/product/i-product';
import {environment} from '../../../environment';
import {HttpClient, HttpParams} from '@angular/common/http';
import {Observable} from 'rxjs';
import { IPage } from '../models/i-page';
import {IProductSearchCriteria} from '../models/i-product-search-criteria';

@Injectable({
  providedIn: 'root',
})
export class ProductService {
  private readonly baseUrl = environment.apiBase + '/products';

  constructor(private http: HttpClient) {}

  createProduct(dto: any, images: File[]): Observable<any> {
    const formData = new FormData();

    const jsonBlob = new Blob([JSON.stringify(dto)], {
      type: 'application/json'
    });

    formData.append('dto', jsonBlob);

    images.forEach(image => {
      formData.append('images', image);
    });

    return this.http.post(`${this.baseUrl}`, formData);
  }

  getProductById(id: string): Observable<IProduct> {
    return this.http.get<IProduct>(`${this.baseUrl}/public/${id}`);
  }

  getProductBySlug(slug: string): Observable<IProduct> {
    return this.http.get<IProduct>(`${this.baseUrl}/public/items/${slug}`);
  }

  getUserProducts(page: number, size: number = 10): Observable<IPage<IProductCard>> {

    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sort', 'createdDate,desc');

    return this.http.get<IPage<IProductCard>>(`${this.baseUrl}/user`, { params });
  }

  deleteProduct(id: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  toggleVisibility(id: string): Observable<void> {
    return this.http.patch<void>(`${this.baseUrl}/${id}/visibility`, {});
  }

  updateProduct(productId: string, dto: any, images: File[]) {
    const formData = new FormData();

    const jsonBlob = new Blob([JSON.stringify(dto)], {
      type: 'application/json'
    });

    formData.append('dto', jsonBlob);

    images.forEach(image => {
      formData.append('newFiles', image);
    });

    return this.http.put(`${this.baseUrl}/${productId}`, formData);
  }

  getFeaturedProducts(page: number, size: number, seed: string): Observable<IPage<IProductCard>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('seed', seed);

    return this.http.get<IPage<IProductCard>>(`${this.baseUrl}/public/featured`, { params });
  }

  search(criteria: IProductSearchCriteria, page: number = 0, size: number = 20): Observable<IPage<IProductCard>> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());

    if (criteria.query) params = params.set('query', criteria.query);
    if (criteria.minPrice) params = params.set('minPrice', criteria.minPrice.toString());
    if (criteria.maxPrice) params = params.set('maxPrice', criteria.maxPrice.toString());
    if (criteria.category) params = params.set('category', criteria.category);
    if (criteria.condition) params = params.set('condition', criteria.condition);
    if (criteria.size) params = params.set('productSize', criteria.size);

    return this.http.get<IPage<IProductCard>>(`${this.baseUrl}/public/search`, { params });
  }

  getSellerProducts(sellerId: string, page: number = 0, size: number = 10): Observable<IPage<IProductCard>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sort', 'createdDate,desc');

    return this.http.get<IPage<IProductCard>>(`${this.baseUrl}/public/seller/${sellerId}`, { params });
  }

  incrementView(id: string): Observable<void> {
    return this.http.patch<void>(`${this.baseUrl}/public/${id}/view`, {});
  }
}
